rs-matter/src/data_model/subscriptions.rs

```diff
@@ -0,0 +1,142 @@
+/*
+ *
+ *    Copyright (c) 2020-2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+use core::cell::RefCell;
+
+use embassy_sync::blocking_mutex::raw::NoopRawMutex;
+use embassy_time::Instant;
+
+use portable_atomic::{AtomicU32, Ordering};
+
+use crate::utils::notification::Notification;
+
+struct Subscription {
+    node_id: u64,
+    id: u32,
+    min_int_secs: u16,
+    max_int_secs: u16,
+    reported_at: Instant,
+    changed: bool,
```

Comment:
`changed` feels a bit vague ... usually I would assume it relates to the underlying structure changed (same as `dirty` ... i.e. reported_at or max_int_secs or such changed).

how about `data_model_dirty` ?
Unsure if implemented: the subscription itself generally may be interested only in a subset of data, so a global `something dirty` likely triggers things more than a granular trigger. This may warrant a comment.

At the same time, for practicality I expect almost any controller except test ones to do a `subscribe *` as their only subscription so as to not use too many resources, so having things granular does not seem as critical.

Reply:
> `changed` feels a bit vague ... usually I would assume it relates to the underlying structure changed (same as `dirty` ... i.e. reported_at or max_int_secs or such changed).
> 
> how about `data_model_dirty` ? 

Maybe `model_changed` or `model_updated`? `dirty` for me implies that you do need to take an extra action to "clean" it or "save" it, like a document that has changes but is not saved yet.

> Unsure if implemented: the subscription itself generally may be interested only in a subset of data, so a global `something dirty` likely triggers things more than a granular trigger. This may warrant a comment.
> 
> At the same time, for practicality I expect almost any controller except test ones to do a `subscribe *` as their only subscription so as to not use too many resources, so having things granular does not seem as critical.

No this is not implemented yet but there are TODOs for this. There are many ways how we can do this, so I left this for later.

Digression:
Precisely to save resources, I was imagining a very simple scheme where - rather than keeping up what exactly had changed (as in which clusters) - the subscription itself - having info what clusters are read by the subscriber in its saved RX buffer - would go though the clusters of all requested attribs and compute a simple md5/shaX hash on their data-vers, and compare it with a previous one. This somehow for me strikes a balance of being too granular, but then YAGNI and just blindly reporting as changed everything.

The above scheme _would_imply that the subscriptions' code only needs to be notified that "something in the model did change", and it would itself reconcile the changes (using its own logic) and decide if to fire or not.